### PR TITLE
Permute schedule for perfectly nested loops

### PR DESCRIPTION
**Pull request by Blealtan**
This PR starts the work on permute loops with customizable iteration space transformation function.
Such schedules are helpful for parallelizing loops with spatially skewed dependencies﻿.

### CODE REVIEW

File: include/schedule/permute.h [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#ifndef FREE_TENSOR_LINEAR_REORDER_H
+#define FREE_TENSOR_LINEAR_REORDER_H
```
Comment (roastduck) [inline include/schedule/permute.h:2]: `FREE_TENSOR_PERMUTE`

File: src/schedule/permute.cc [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+            return inner;
+        } else {
+            return Mutator::visit(op);
+        }
+    }
+
+    Expr visit(const Var &op) override {
+        // lookup the map; if exists, replace iter with provided expression
+        if (iterReplacer_.find(op->name_) != iterReplacer_.end())
+            return iterReplacer_[op->name_];
```
Comment (roastduck) [inline src/schedule/permute.cc:81]: Better C++:

```
if (auto it = iterReplacer_.find(op->name_); it != iterReplacer_.end())
    return it->second;
```

File: src/schedule/permute.cc [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+            if (loopsId[i] != item->id())
+                error("should be in the same order as the input");
+    }
+
+    // prepare the permutation map
+    std::string permuteMapStr;
+    {
+        auto originalIter =
+            iter::range(loops.size()) | iter::imap([](auto &&i) {
+                return makeVar("din" + std::to_string(i))
```
Comment (roastduck) [inline src/schedule/permute.cc:127]: Do you mean "dim"?

Comment (roastduck) [reply src/schedule/permute.cc:127]: Ok, I see it's d in

**Comment by Blealtan** [comment]
Need to add some documents, not merging now

**Review by roastduck — APPROVED** [review]

**Pull request merged**